Geoblocking ADR draft (#1309)
kyma-project/api-gateway: docs/contributor/adr/0010-geoblocking.md

# Geoblocking

## Status

Accepted

## Context

Geoblocking is a feature allowing to block incoming traffic coming from certain IP addresses, that are exclusive to certain countries or regions.
Can be utilized against anonymous users and system to system network communication, where both are identified only by its source IP address.
Many companies are using geoblocking to protect their services from unwanted traffic, such as DDoS attacks, or to comply with legal regulations.

> Therefore, creating this kind of feature is a convenience for the user, as it extracts the need to implement it on its own. 

Reviewer: We (try to) use camel case for Kubernetes resources' kinds in our docs like Deployments, Pods, Services, etc. So, I also used "GeoBlocking" when we talk about the CR or the name of the controller/operator but left "geoblocking" when it's about the feature.

- version information - used in events
- etag - used when checking for updates
- lastUpdateCheckTime - the time when the list update has been performed (successfully) 
- lastUpdateTime - the time when the list has been updated

The above Config Map should only be used by the ip-auth service for caching/fallback purpose, it can't be treated as an official contract. It may be replaced with some other solution at any time without further notice.

Example of an IP range allow/block list cache:

```
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: ip-range-list-cache
data:
  version: ip-list-20240912-2200
  etag: PeJwSe7ncNQchtGbkCPJ
  lastUpdateCheckTime: "2024-09-17T15:28:50Z"
  lastUpdateTime: "2024-09-17T14:28:50Z"
  allow: |
    - 192.0.2.10/32
    - 2001:0002:6c::430
  block: |
    - 192.0.2.0/24
    - 198.51.100.0/24
    - 203.0.113.0/24
    - 2001:0002::/48
```

![SAP internal service](../../assets/geoblocking-SAP-service.drawio.svg)

### Geoblocking CR and controller

In order to ensure reliability and configurability a new Geoblocking Custom Resource and a new Geoblocking Controller is introduced. The controller is responsible for:

> - managing the ip-auth service deployment

Reviewer: ```suggestion
- managing the ip-auth Service's Deployment
```

- managing authorization policy that plugs ip-auth authorizer to all incoming requests
- performing configuration checks (like external traffic policy)
- reporting geoblocking state

> There should be only one Geoblocking resource. The controller should set the additional Geoblocking CR in `Error` state, similarly to [APIGateway CR](https://github.com/kyma-project/api-gateway/blob/05ed57f3299f42c2565ed1b28b84dee808a5213b/controllers/operator/apigateway_controller.go#L100).

Reviewer: ```suggestion
There should be only one GeoBlocking resource. The controller should set the additional Geoblocking CR in the `Error` state, similarly to [APIGateway CR](https://github.com/kyma-project/api-gateway/blob/05ed57f3299f42c2565ed1b28b84dee808a5213b/controllers/operator/apigateway_controller.go#L100).
```

![Geoblocking controller](../../assets/geoblocking-cr-controller.drawio.svg)

#### CR examples

##### Usage of SAP internal service

```yaml
apiVersion: geoblocking.kyma-project.io/v1alpha1
kind: GeoBlocking
metadata:
  name: geoblocking
  namespace: kyma-system
spec:
  gbService:
    secret: "gb-secret"
    tokenUrl: https://oauth.example.com/oauth2/token
    ipRange:
      url: https://lists.example.com/lists/some-list
      refreshInSeconds: 3600
    events:
      url: https://events.example.com/events
      lobId: "kyma"
      productId: "some-product"
      systemId: "some-system"
```

##### IP range allow/block list populated by customer
```yaml
apiVersion: geoblocking.kyma-project.io/v1alpha1
kind: GeoBlocking
metadata:
    name: geoblocking
    namespace: kyma-system
spec:
  ipRange:
    configMap: "my-own-ip-range-list"
```

### Technical details

#### Policy download optimization

> In order to reduce unnecessary updates of the list of blocked IP ranges the ip-auth service should use ETag mechanism, which is supported by the SAP internal service.

Reviewer: ```suggestion
In order to reduce unnecessary updates to the list of blocked IP ranges, the ip-auth Service should use the ETag mechanism, which is supported by the SAP internal service.
```

This includes at least the following details:
- deployment:
  - replicas / scaling
  - update strategy
  - Docker image
  - resources (limits, requests)
  - probes (liveness, readiness)
  - exposed ports
  - security context (user, etc.)
  - service account
  - configmap mount (ip allow / block list)
  - secret mount (secrets for Geoblocking service)
- service:
  - selectors
  - ports

#### Authorization Policy settings

For simplicity reasons, we assume that geoblocking would be applied globally for the whole application.

This means that it just applies to the whole Istio Ingress Gateway and there is no need expose any additional configuration in this scope.

For now we assume that the feature works with default Istio Ingress Gateway installed by the Istio module.

#### Namespaces

Geoblocking resources would be placed in the following namespaces:

| Resource                                    | Who creates it                 | Namespace                          |
| ------------------------------------------- | ------------------------------ | ---------------------------------- |

> | Geoblocking Operator (API Gateway Operator) | Power user / Lifecycle Manager | kyma-system                        |
> | Geoblocking CR                              | Power user                     | kyma-system                        |
> | ip-auth Deployment                          | Geoblocking Operator           | kyma-system                        |
> | ip-auth Service                             | Geoblocking Operator           | kyma-system                        |
> | ip-auth Secret                              | Power user                     | kyma-system                        |
> | IP ranges Config Map with input             | ip-auth or External customer   | kyma-system                        |
> | IP ranges Config Map with cache             | ip-auth or External customer   | kyma-system                        |
> | Authorization Policy                        | Geoblocking Operator           | istio-system                       |

Reviewer: ```suggestion
| GeoBlocking Operator (API Gateway Operator) | Power user / Lifecycle Manager | `kyma-system`                        |
| GeoBlocking CR                              | Power user                     | `kyma-system`                        |
| ip-auth Deployment                          | GeoBlocking Operator           | `kyma-system`                        |
| ip-auth Service                             | GeoBlocking Operator           | `kyma-system`                        |
| ip-auth Secret                              | Power user                     | `kyma-system`                        |
| IP ranges ConfigMap with input             | ip-auth or external customer   | `kyma-system`                        |
| IP ranges ConfigMap with cache             | ip-auth or external customer   | `kyma-system`                        |
| AuthorizationPolicy                        | GeoBlocking Operator           | `istio-system`                       |
```

```yaml
spec:
  gbService:
    events:
      queue:
        size: 100
        ttl: 360
  deployment: 
    resources:
      requests:
        memory: "64Mi"
        cpu: "250m"
      limits:
        memory: "128Mi"
        cpu: "500m"
```

Decision: it doesn't make sense to expose so many technical parameters, especially that there would be no easy way to determine the above values. It is better to hardcode these parameters with some reasonable values (for now), potentially providing multiple variants (evaluation and production profiles).

### Configurability of the rules in Authorization Policy

Istio allows to specify a lot of conditions that are used to link a request with a given Authorization Policy: https://istio.io/latest/docs/reference/config/security/authorization-policy/#Rule

We can either:
- expose them and make it configurable in the CR
- hide it and doesn't allow such configurability

Decision: it doesn't make sense to expose so many detailed parameters, because the idea behind Geoblocking is to be compliant with regulations, which are usually very global (apply to the whole company / product / application).

### IP range allow/block list fallback/caching

> The list of IP ranges is processed by each ip-auth container and stored in its memory and regularly refreshed using geoblocking service. If geoblocking service doesn't work then the ip-auth containers may work as before, the list is not refreshed.

Author: GeoBlocking Service replaced with the SAP internal service

> However, the problem is when the ip-auth container needs to be restarted when geoblocking service is not available, so it can't download the list of IP ranges. This may be the case during an upgrade of ip-auth or in the event of runtime problems (for example ouf of memory). It is not acceptable to allow the traffic in such case. This means that the list of IP ranges should be stored somehow to be able to survive the pod restarts. This requirement is refered to as 'fallback'.

Author: GeoBlocking Service replaced with the SAP internal service

Decision: Hardcode deployment details in the API Gateway module.

Consequence: It won't be possible to change the deployment details. 

### Propagation of ip-auth service issues

Geoblocking controller is responsible for the reconciliation of the Geoblocking custom resource, in particular: creating deployment of ip-auth service, creation of Authorization Policy, doing some configuration checks, etc. It may report issues related to above actions in the Geoblocking custom resource status.

However, some issues may be visible only in the ip-auth service. The good example is a problem with the communication with SAP internal service (like a wrong secret).

There is no easy way to propagate such issues to the Geoblocking CR. This would require a dedicated mechanism - like a 'health' endpoint exposed by ip-auth containers, so it can be analyzed by the Geoblocking controller.

Decision: Don't implement additional mechanisms and follow the 'standard' approach, so the controller reports issues in the CR only from its own layer, while the ip-auth service reports issues in its log or via liveness/readiness probes (like every other workload).

> Consequence: Geoblocking CR won't be responsible for presenting Geoblocking 'health'. Support or developers must be aware that they should also inspect ip-auth logs in case of issues.

Author: We discussed it, should be geoblocking health

### Namespaces

Geoblocking mechanism touches multiple resources in multiple areas:
- Istio configuration

> - Ingress gateway

Reviewer: ```suggestion
- Ingress Gateway
```

- Gateways in general
- Authorization Policy

> - Authorizer workload (service, deployment)

Reviewer: ```suggestion
- Authorizer workload (Service, Deployment)
```

> - Configuration of SAP internal service

Reviewer: ```suggestion
- Configuration of the SAP internal service
```
